packages/env-tests/src/tests/granda-mento.ts

```diff
@@ -0,0 +1,146 @@
+
+            const grandaMento = await context.kit.contracts.getGrandaMento()
+
+            let sellToken
+            let sellAmount
+            if (sellCelo) {
+              sellToken = await context.kit.contracts.getGoldToken()
+              sellAmount = celoAmountToSell
+            } else {
+              sellToken = await context.kit.celoTokens.getWrapper(stableToken as StableToken)
+              sellAmount = stableTokenAmountToSell
+            }
+            await sellToken
+              .approve(grandaMento.address, sellAmount.toFixed())
+              .sendAndWaitForReceipt({
+                from: from.address,
+              })
+            logger.debug(
+              {
+                sellAmount,
+                sellTokenStr,
+                spender: grandaMento.address,
+              },
+              'Approved GrandaMento to spend sell token'
+            )
+
+            // Some flakiness has been observed after approving, so we sleep
+            await sleep(5000)
+
+            const sellTokenBalanceBeforeCreation = await sellToken.balanceOf(from.address)
+
+            const creationTx = await grandaMento.createExchangeProposal(
+              context.kit.celoTokens.getContract(stableToken as StableToken),
+              sellAmount,
+              sellCelo
+            )
+            const creationReceipt = await creationTx.sendAndWaitForReceipt({
+              from: from.address,
+            })
+            const minedCreationTx = await context.kit.web3.eth.getTransaction(
+              creationReceipt.transactionHash
+            )
+            const proposalId = creationReceipt.events!.ExchangeProposalCreated.returnValues
+              .proposalId
+
+            logger.debug(
+              {
+                sellAmount,
+                sellCelo,
+                proposalId,
+              },
+              'Created exchange proposal'
+            )
+
+            let celoFees = new BigNumber(creationReceipt.gasUsed).times(minedCreationTx.gasPrice)
+
+            const sellTokenBalanceAfterCreation = await sellToken.balanceOf(from.address)
+
+            // If we are looking at the CELO balance, take the fees spent into consideration.
+            const expectedBalanceDifference = sellCelo ? sellAmount.plus(celoFees) : sellAmount
+
+            expect(
+              sellTokenBalanceBeforeCreation.minus(sellTokenBalanceAfterCreation).toString()
+            ).toBe(expectedBalanceDifference.toString())
+
+            const cancelReceipt = await grandaMento
+              .cancelExchangeProposal(proposalId)
+              .sendAndWaitForReceipt({
+                from: from.address,
+              })
+            const minedCancelTx = await context.kit.web3.eth.getTransaction(
+              cancelReceipt.transactionHash
+            )
+
+            logger.debug(
+              {
+                proposalId,
+              },
+              'Cancelled exchange proposal'
```

Comment:
Should there also be a test for the exchange being approved and completed?

Caveat: I'm not 100% clear on the scope and limitations of env-tests.

Reply:
Hmm I've been flip-flopping on a couple other test cases because they require waiting long periods of time and I'm not sure if env-tests are well suited for that.
* Approving & governance vetoing - passing a governance proposal to veto requires waiting between stages
* Approving & executing - requires waiting the veto period before the proposal is able to execute

I'll chat with Nam to see what his take is given he probably has the most context on env-tests, but I think these may need to be manually tested on a testnet

Reply:
Ended up adding that test case